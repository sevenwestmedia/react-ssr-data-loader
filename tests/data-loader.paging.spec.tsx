import { PagedComponentFixture } from './helpers/paged-component-fixture'
import Adapter from 'enzyme-adapter-react-16'
import { configure } from 'enzyme'
import { act } from 'react-dom/test-utils'
import { processEventLoop } from './helpers/event-loop-helpers'

configure({ adapter: new Adapter() })
describe('data-loader', () => {
    it('supports paged data', async () => {
        const sut = new PagedComponentFixture(undefined, 'testKey', false)

        await act(() => sut.testDataPromise.resolve(['Test']))
        sut.assertState()

        act(() => sut.nextPage())

        sut.assertState()

        await act(() => sut.testDataPromise.resolve(['Test2']))
        sut.assertState()
    })

    it('paged component supports being re-rendered', async () => {
        const sut = new PagedComponentFixture(undefined, 'testKey', false)

        await act(async () => {
            await sut.testDataPromise.resolve(['Test'])
            sut.nextPage()
            await sut.testDataPromise.resolve(['Test2'])

            sut.root.setProps({ id: 'testKey' })
        })

        sut.assertState()

        await act(() => processEventLoop())

        sut.assertState()
    })

    it('calling nextPage when already loading a page ignores the action', async () => {
        const sut = new PagedComponentFixture(undefined, 'testKey', false)

        await act(async () => {
            await sut.testDataPromise.resolve(['Test'])
            sut.nextPage()
            sut.nextPage()
        })

        sut.assertState()
    })

    it('refresh causes page to go back to page 1', async () => {
        const sut = new PagedComponentFixture(undefined, 'testKey', false)

        await act(async () => {
            await sut.testDataPromise.resolve(['Test'])
            sut.nextPage()
            await sut.testDataPromise.resolve(['Test2'])
            sut.refreshData()
            await sut.testDataPromise.resolve(['Test'])
        })

        sut.assertState()
    })
})
