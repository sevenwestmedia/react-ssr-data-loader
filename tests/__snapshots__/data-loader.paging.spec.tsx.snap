// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data-loader calling nextPage when already loading a page ignores the action 1`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 2,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 3,
  "renderParams": Object {
    "id": "testKey",
    "page": 2,
    "paging": Object {
      "pageSize": 10,
    },
  },
  "renderProps": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Fetching",
  },
}
`;

exports[`data-loader paged component supports being re-rendered 1`] = `
Object {
  "loadAllCompletedCalled": 2,
  "loadDataCount": 2,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 4,
  "renderParams": Object {
    "id": "testKey",
    "page": 2,
    "paging": Object {
      "pageSize": 10,
    },
  },
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Array [
        "Test2",
      ],
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`data-loader paged component supports being re-rendered 2`] = `
Object {
  "loadAllCompletedCalled": 2,
  "loadDataCount": 2,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 4,
  "renderParams": Object {
    "id": "testKey",
    "page": 2,
    "paging": Object {
      "pageSize": 10,
    },
  },
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Array [
        "Test2",
      ],
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`data-loader refresh causes page to go back to page 1 1`] = `
Object {
  "loadAllCompletedCalled": 3,
  "loadDataCount": 3,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 6,
  "renderParams": Object {
    "id": "testKey",
    "page": 2,
    "paging": Object {
      "pageSize": 10,
    },
  },
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Array [
        "Test",
      ],
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`data-loader supports paged data 1`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 1,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 2,
  "renderParams": Object {
    "id": "testKey",
    "page": 1,
    "paging": Object {
      "pageSize": 10,
    },
  },
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Array [
        "Test",
      ],
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;

exports[`data-loader supports paged data 2`] = `
Object {
  "loadAllCompletedCalled": 1,
  "loadDataCount": 2,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 3,
  "renderParams": Object {
    "id": "testKey",
    "page": 2,
    "paging": Object {
      "pageSize": 10,
    },
  },
  "renderProps": Object {
    "data": Object {
      "hasData": false,
    },
    "lastAction": Object {
      "success": true,
      "type": "none",
    },
    "status": "Fetching",
  },
}
`;

exports[`data-loader supports paged data 3`] = `
Object {
  "loadAllCompletedCalled": 2,
  "loadDataCount": 2,
  "renderActions": Object {
    "nextPage": [Function],
    "refresh": [Function],
  },
  "renderCount": 4,
  "renderParams": Object {
    "id": "testKey",
    "page": 2,
    "paging": Object {
      "pageSize": 10,
    },
  },
  "renderProps": Object {
    "data": Object {
      "dataFromServerSideRender": false,
      "hasData": true,
      "result": Array [
        "Test2",
      ],
    },
    "lastAction": Object {
      "success": true,
      "type": "fetch",
    },
    "status": "Idle",
  },
}
`;
